lib/client/escape/reader.go

```diff
@@ -0,0 +1,187 @@
+	escapeDisconnect = "\r~."
+	// Note: on a raw terminal, "\r\n" is needed to move a cursor to the start
+	// of next line.
+	helpText = "\r\ntsh escape characters:\r\n  ~? - display a list of escape characters\r\n  ~. - disconnect\r\n"
+)
+
+var (
+	// ErrDisconnect is returned when the user has entered a disconnect
+	// sequence, requesting connection to be interrupted.
+	ErrDisconnect = errors.New("disconnect escape sequence detected")
+	// ErrTooMuchBufferedData is returned when the Reader's internal buffer has
+	// filled over 10MB. Either the consumer of Reader can't keep up with the
+	// data or it's entirely stuck and not consuming the data.
+	ErrTooMuchBufferedData = errors.New("internal buffer has grown too big")
+)
+
+// Reader is an io.Reader wrapper that catches OpenSSH-like escape sequences in
+// the input stream. See NewReader for more info.
+//
+// Reader is safe for concurrent use.
+type Reader struct {
+	inner        io.Reader
+	out          io.Writer
+	onDisconnect func(error)
+	bufferLimit  int
+
+	// cond protects buf and err and also announces to blocked readers that
+	// more data is available.
+	cond sync.Cond
+	buf  []byte
+	err  error
+}
+
+// NewReader creates a new Reader to catch escape sequences from 'in'.
+//
+// Two sequences are supported:
+// - "~?": prints help text to 'out' listing supported sequences
+// - "~.": disconnect stops any future reads from in; after this sequence,
+//   callers can still read any unread data up to this sequence from Reader but
+//   all future Read calls will return ErrDisconnect; onDisconnect will also be
+//   called with ErrDisconnect immediately
+//
+// NewReader starts consuming 'in' immediately in the background. This allows
+// Reader to detect sequences without the caller actively calling Read (such as
+// when it's stuck writing out the received data).
+//
+// Unread data is accumulated in an internal buffer. If this buffer grows to a
+// limit (currently 10MB), Reader will stop permanently. onDisconnect will get
+// called with ErrTooMuchBufferedData. Read can still be called to consume the
+// internal buffer but all future reads after that will return
+// ErrTooMuchBufferedData.
+//
+// If the internal buffer is empty, calls to Read will block until some data is
+// available or an error occurs.
+func NewReader(in io.Reader, out io.Writer, onDisconnect func(error)) *Reader {
+	r := newUnstartedReader(in, out, onDisconnect)
+	go r.runReads()
+	return r
+}
+
+// newUnstartedReader allows unit tests to mutate Reader before runReads
+// starts.
+func newUnstartedReader(in io.Reader, out io.Writer, onDisconnect func(error)) *Reader {
+	return &Reader{
+		inner:        in,
+		out:          out,
+		onDisconnect: onDisconnect,
+		bufferLimit:  readerBufferLimit,
+		cond:         sync.Cond{L: &sync.Mutex{}},
+		// note: no need to pre-allocate buf, it will allocate and grow as
+		// needed in runReads via append.
+	}
+}
+
+func (r *Reader) runReads() {
```

Comment:
Few notes based on comparing this function to the [`process_escapes`](https://github.com/openssh/openssh-portable/blob/3ab6fccc3935e9b778ff52f9c8d40f215d58e01d/clientloop.c#L937) function in `openssh`:

- Help escape sequence is not filtered out (`openssh` appears to filter out the help sequence).
- Disconnect escape sequence drops all preceding bytes (`openssh` appears to send all date *up to* the disconnect escape sequence).
- `\r~` starts an escape, but `\n~` does not (`openssh` treats `\r` and `\n` interchangeably here).
- Reads ending in partial escape sequence (e.g. `\n~`) will send the escape character (`~`) to remote (`openssh` does not send the escape character until the following character is available to ensure that partial escape sequences are not transmitted to remote).

I think its preferable if we keep parity with `openssh` on these points (potentially confusing if we don't, since we shadow their syntax).

Reply:
Also, the escape character should probably be configurable (or at least the behavior should be able to be deactivated).  I don't know what this might break, but its probably best not to roll out a behavioral change like this without the ability to work around it if it does break something.

Reply:
I think @fspmarshall covered this in the fourth bullet, but this will not correctly recognize sequences that are split between buffers. Processing character-by-character is safer.

Reply:
Just a meta-response: I agree that it's not a perfect replica of openssh behavior. But it works for the use-case it's addressing (killing stuck sessions).

That being said, I'll massage the code some more to prevent escape sequences from going out to the remote side.

> Disconnect escape sequence drops all preceding bytes (openssh appears to send all data up to the disconnect escape sequence).

What does it do when remote side is unresponsive?

> `\r~ starts an escape, but \n~ does not (openssh treats \r and \n interchangeably here).`

Nice catch, will fix!

> Also, the escape character should probably be configurable (or at least the behavior should be able to be deactivated).

This makes sense, I'll add a tsh flag.
Although I'd like to not over-complicate this niche feature.

>  this will not correctly recognize sequences that are split between buffers.

It will _recognize_ sequences split between buffers via the `prev` buffer. But it will not _block_ them the same way as openssh does.

Reply:
> What does it do when remote side is unresponsive?

Ah, good point.  Take this with a grain of salt because my knowledge of both `openssh` and `C` in general is pretty weak, and the state-machine around channels in `openssh` is complex.... but I believe it performs a single non-blocking write attempt with any remaining pre-escape sequence data and then exits.  Thats a much weaker guarantee than I originally assumed, so I'd say forget that suggestion and keep dropping the preceeding bytes.

Reply:
Ack. I don't see an easy way to do a non-blocking write at our level of abstraction in `lib/client/session.go` so I'll keep it as is.

Refactored the reader logic to read one character at a time and filter out escape sequences.
Also added handling for both `\r` and `\n`.

Next is the tsh flag plumbing...

Reply:
Added `--no-enable-escape-sequences` flag to tsh.
PTAL